src/Avalonia.Base/Media/TextFormatting/ShapedTextRun.cs

```diff
@@ -195,10 +184,7 @@ internal GlyphRun CreateGlyphRun()
                 ShapedBuffer.GlyphTypeface,
```

Comment:
You can make GlyphRun just out of this ShapedBuffer. Other parameters are redundant because already inside ShapedBuffer.

Reply:
Sorry, I'm not sure I'm following here? If you refer to the `ShapedBuffer` argument that's the `GlyphInfo` list: I replaced the four arrays with one.

Reply:
nevermind. I've carefully merged your commit to my code and missed some differenced that I made.
To use even less memory I re-designed GlyphRun to be(in my private repo):
```
public sealed class GlyphRun : IDisposable
{
    public readonly ShapedBuffer Buffer;

    public GlyphRun(ShapedBuffer buffer) => Buffer = buffer;
```
Maybe there is an option to do the same in Avalonia, as I did for myself, but I think it's an API change and will not be accepted, so sorry for making some confusion here.

Reply:
Thank you for the clarification!

With the changes it's almost equivalent to passing the `ShapedBuffer`: there are no extra allocations. Removing the fields (rendering size, bidi level, etc.) from the `GlyphRun` to access the `ShapedBuffer` instead would only save a few primitive fields in the `GlyphRun`, which are negligible.

That said, I also feel like the `ShapedBuffer` and `GlyphRun` implementations could probably be merged (I'd remove `ShapedBuffer`): `ShapedBuffer` passes everything to `GlyphRun`, and both classes have some similar methods such as `FindGlyphByIndex` (with slightly different implementations).